sdk/storage/Azure.Storage.Blobs/src/PageBlobWriteStream.cs

```diff
@@ -0,0 +1,144 @@
+
+namespace Azure.Storage.Blobs
+{
+    internal class PageBlobWriteStream : StorageWriteStream
+    {
+        private readonly PageBlobClient _pageBlobClient;
+        private readonly PageBlobRequestConditions _conditions;
+        private long _writeIndex;
+
+        public PageBlobWriteStream(
+            PageBlobClient pageBlobClient,
+            long bufferSize,
+            long position,
+            PageBlobRequestConditions conditions,
+            IProgress<long> progressHandler) : base(
+                position,
+                bufferSize,
+                progressHandler)
+        {
+            ValidateBufferSize(bufferSize);
+            ValidatePosition(position);
+            _pageBlobClient = pageBlobClient;
+            _conditions = conditions ?? new PageBlobRequestConditions();
+            _writeIndex = position;
+        }
+
+        protected override async Task WriteInternal(
+            byte[] buffer,
+            int offset,
+            int count,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            ValidateWriteParameters(buffer, offset, count);
+            int remaining = count;
+
+            // New bytes will fit in the buffer.
+            if (count <= _bufferSize - _buffer.Position)
+            {
+                await WriteToBufferInternal(buffer, offset, count, async, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // We need a multiple of 512 to flush.
+                if (_buffer.Length % Constants.Blob.Page.PageSizeBytes != 0)
+                {
+                    int bytesToWrite = (int)(Constants.Blob.Page.PageSizeBytes - _buffer.Length % Constants.Blob.Page.PageSizeBytes);
+                    await WriteToBufferInternal(buffer, offset, bytesToWrite, async, cancellationToken).ConfigureAwait(false);
+                    remaining -= bytesToWrite;
+                    offset += bytesToWrite;
+                }
+
+                // Flush the buffer.
+                await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+
+                while (remaining > 0)
+                {
+                    await WriteToBufferInternal(
+                        buffer,
+                        offset,
+                        (int)Math.Min(remaining, _bufferSize),
+                        async,
+                        cancellationToken).ConfigureAwait(false);
+
+                    // Remaining bytes won't fit in buffer.
+                    if (remaining > _bufferSize)
+                    {
+                        await AppendInternal(async, cancellationToken).ConfigureAwait(false);
+                        remaining -= (int)_bufferSize;
+                        offset += (int)_bufferSize;
+                    }
+
+                    // Remaining bytes will fit in buffer.
+                    else
+                    {
+                        remaining = 0;
+                    }
```

Comment:
No, because the number of bytes we are writing to the buffer is `(int)Math.Min(remaining, _bufferSize),`.  In the case `remaining > _bufferSize`, we just wrote `_bufferSize` to the buffer, which is an increment of 512.